lib/src/rules/voidness_propagation.dart

```diff
@@ -0,0 +1,165 @@
+    _check(type, node.rightHandSide?.staticType, node.rightHandSide);
+  }
+
+  @override
+  void visitExpressionFunctionBody(ExpressionFunctionBody node) {
+    DartType returnType;
+    if (node?.parent?.parent is FunctionDeclaration) {
+      returnType = (node.parent.parent as FunctionDeclaration).returnType?.type;
+    } else if (node?.parent is MethodDeclaration) {
+      returnType = (node.parent as MethodDeclaration).returnType?.type;
+    }
+    _check(returnType, node.expression.staticType, node.expression);
+  }
+
+  @override
+  void visitInstanceCreationExpression(InstanceCreationExpression node) {
+    final args = node.argumentList.arguments;
+    final parameters = node.staticElement?.parameters;
+    if (parameters != null) {
+      _checkArgs(args, parameters);
+    }
+  }
+
+  @override
+  void visitMethodDeclaration(MethodDeclaration node) {
+    final inheritedMethod = DartTypeUtilities.lookUpInheritedMethod(node);
+    if (inheritedMethod != null) {
+      _check(
+          inheritedMethod.returnType, node.returnType?.type, node.returnType);
+    } else {
+      for (final e in DartTypeUtilities.overridenMethods(node)) {
+        _check(e.returnType, node.returnType?.type, node.returnType);
+      }
+    }
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final type = node.staticInvokeType;
+    if (type is FunctionType) {
+      final args = node.argumentList.arguments;
+      final parameters = type.parameters;
+      _checkArgs(args, parameters);
+    }
+  }
+
+  @override
+  void visitReturnStatement(ReturnStatement node) {
+    final parent = node.getAncestor((e) =>
+        e is FunctionExpression ||
+        e is MethodDeclaration ||
+        e is FunctionDeclaration);
+    if (parent is FunctionExpression) {
+      final type = parent.staticType;
+      if (type is FunctionType) {
+        _check(type.returnType, node.expression?.staticType, node.expression);
+      }
+    } else if (parent is MethodDeclaration) {
+      _check(parent.declaredElement.returnType, node.expression?.staticType,
+          node.expression);
+    } else if (parent is FunctionDeclaration) {
+      _check(parent.declaredElement.returnType, node.expression?.staticType,
+          node.expression);
+    }
+  }
+
+  @override
+  void visitVariableDeclaration(VariableDeclaration node) {
+    if (node.initializer != null) {
+      _check((node.parent as VariableDeclarationList).type?.type,
+          node.initializer.staticType, node.initializer);
+    }
+  }
+
+  void _check(DartType expectedType, DartType type, AstNode node) {
+    if (expectedType == null || type == null) {
+      return;
+    } else if (expectedType is FunctionType && type is FunctionType) {
+      _check(expectedType.returnType, type.returnType, node);
+    } else if (type is InterfaceType &&
+        expectedType is InterfaceType &&
+        type != expectedType &&
+        type.element == expectedType.element) {
```

Comment:
Hmm, this won't catch `Iterable<void>` being assigned to `List<int>`, or `List<int>` being assigned to `Iterable<void>`.

There are other pathological cases too:

```
abstract class IV implements Iterable<void> {}
abstract class LI implements List<int> {}

void bad(IV iv, LI li) {
  li = iv;
}
```

I guess this is probably the most complex part and the API change I described earlier wouldn't help you solve this...once again, though, this is logic we have implemented in the analyzer. I wonder if we can expose that API to you at all in a way where you can provide custom hooks into the subtype comparison. FWIW the subtype compare code is here: https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/src/generated/type_system.dart#L1480

Reply:
We actually _do_ have a public API for subtype comparison: https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/dart/element/type.dart#L154.